Moonscraper Chart Editor/Assets/Scripts/Game/Charts/IO/Midi/MidReader.cs

```diff
@@ -506,37 +509,74 @@ private static void ReadNotes(IList<MidiEvent> track, IList<MessageProcessParams
                     }
                 }
 
-                var note = track[i] as NoteOnEvent;
-                if (note != null && note.OffEvent != null)
+                if (trackEvent is NoteOnEvent noteOn)
                 {
+                    // Check for duplicates
+                    // (a corresponding note off must be found, two note ons that are the same are disallowed)
+                    bool duplicate = false;
+                    foreach (var (note, _) in noteQueue)
+                    {
+                        if (note.NoteNumber == noteOn.NoteNumber && note.Channel == noteOn.Channel)
+                        {
```

Comment:
Try to avoid foreach iterator allocation within potentially high density loops like these.
```suggestion
                    for (int noteEventQueueIndex = 0; ...
                    {
                        var note = noteQueue[noteEventQueueIndex].note;
                        if (note.NoteNumber == noteOn.NoteNumber && note.Channel == noteOn.Channel)
                        {
```